pennylane/gradients/pulse_gradient.py

```diff
@@ -564,23 +568,29 @@ def ansatz(params):
     return _expval_stoch_pulse_grad(tape, argnum, num_split_times, key, shots, use_broadcasting)
 
 
-def _generate_tapes_and_cjacs(tape, idx, key, num_split_times, use_broadcasting):
+def _generate_tapes_and_cjacs(tape, op_id, key, num_split_times, use_broadcasting, par_idx=None):
     """Generate the tapes and compute the classical Jacobians for one given
     generating Hamiltonian term of one pulse.
     """
-    op, op_idx, term_idx = tape.get_operation(idx)
-    if not isinstance(op, ParametrizedEvolution):
-        raise ValueError(
-            "stoch_pulse_grad does not support differentiating parameters of "
-            "other operations than pulses."
-        )
-
+    op, op_idx, term_idx = op_id
     coeff, ob = op.H.coeffs_parametrized[term_idx], op.H.ops_parametrized[term_idx]
-    cjac_fn = jax.jacobian(coeff, argnums=0)
+    if par_idx is None:
+        cjac_fn = jax.jacobian(coeff, argnums=0)
+    else:
+        # For `par_idx is not None`, we need to extract the entry of the coefficient
+        # Jacobian that belongs to the parameter of interest. This only happens when
+        # more than one parameter effectively feeds into one coefficient (HardwareHamiltonian)
+
+        def cjac_fn(params, t):
+            return jax.jacobian(coeff, argnums=0)(params, t)[par_idx]
```

Comment:
I'm not entirely sure I understand everything here, but it sounds like for the non-HardwareHamiltonian case the transform assumes that coefficients and parameters map one to one, i.e. that $\partial f_i / \partial \theta_j = \delta_{ij}$ - is that correct? 

**If yes**, That may be problematic because this assumption is not made explicit anywhere. E.g. in the [ctrl-vqe demo](https://pennylane.ai/qml/demos/tutorial_pulse_programming101.html) I did a qnode that looked like

```python
@qml.qnode(dev, interface="jax")
def qnode(theta, t=duration):
    qml.BasisState(list(data.tapered_hf_state), wires=H_obj.wires)
    qml.evolve(H_pulse)(params=(*theta, *theta), t=t)
    return qml.expval(H_obj)
```

A mwe would be:
```python
H = jnp.polyval * qml.PauliX(0) + jnp.polyval * qml.PauliZ(0)

dev = qml.device("default.qubit.jax", wires=1)
@qml.qnode(dev, interface="jax")
def qnode(p):
    qml.evolve(H)((p, p), 10.) # parameters shared
    return qml.expval(qml.PauliZ(0))

p = jnp.ones(2)
jax.grad(qnode)(p)
```

Reply:
This should be fine, because the parameter sharing is on the QNode level. On the tape level (where the transform acts), there is no parameter sharing going on.

For non-hardware Hamiltonians we indeed assume that a tape-level parameter only feeds into one coefficient function. This should be guaranteed by the `ParametrizedEvolution` abstraction, which just is broken by `HardwareHamiltonian`.

If I do `qml.evolve(H)(params, t)`, the `params` are handed to the `ParametrizedHamiltonian` `H`, which sends one entry to each parametrized coefficient, right?

Reply:
The MWE from above:
```python
H = jnp.polyval * qml.PauliX(0) + jnp.polyval * qml.PauliZ(0)

dev = qml.device("default.qubit.jax", wires=1)
@qml.qnode(dev, interface="jax", diff_method=qml.gradients.stoch_pulse_grad, num_split_times=100, use_broadcasting=True)
def qnode(p):
    qml.evolve(H)((p, p), 10.) # parameters shared
    return qml.expval(qml.PauliZ(0))
@qml.qnode(dev, interface="jax")
def qnode_jax(p):
    qml.evolve(H)((p, p), 10.) # parameters shared
    return qml.expval(qml.PauliZ(0))

p = jnp.ones(2)
```
```pycon
>>> jax.grad(qnode)(p)
[-4.03796446 -0.84172773]
>>> jax.grad(qnode_jax)(p)
[-4.21359077 -0.84271786]
```